// opensky-api/build/module/lib/OpenSkyApi.js
('use strict');
import axios from 'axios';
import { URLSearchParams } from 'iso-url';
import { stateVectorMapper } from '../mappers/StateVectorMapper';
import { flightMapper } from './../mappers/FlightMapper';
const axiosConfig = {
    timeout: 5000,
    headers: { 'User-Agent': 'OpenSkyApi-TS/1.0' },
};
export class OpenSkyApi {
    static HOST = 'opensky-network.org';
    static API_ROOT = `https://${this.HOST}/api`;
    static STATES_URI = `${this.API_ROOT}/states/all`;
    static MY_STATES_URI = `${this.API_ROOT}/states/own`;
    static FLIGHTS_URI = `${OpenSkyApi.API_ROOT}/flights/all`;
    static FLIGHTS_BY_AIRCRAFT_URI = `${OpenSkyApi.API_ROOT}/flights/aircraft`;
    static FLIGHTS_BY_ARRIVAL_URI = `${OpenSkyApi.API_ROOT}/flights/arrival`;
    static FLIGHTS_BY_DEPARTURE_URI = `${OpenSkyApi.API_ROOT}/flights/departure`;
    _axios;
    authenticated = false;
    lastRequestTime = {
        GET_STATES: null,
        GET_MY_STATES: null,
        GET_FLIGHTS: null,
    };
    constructor(credentials) {
        if (credentials?.username && credentials?.password) {
            axiosConfig.auth = {
                username: credentials?.username,
                password: credentials?.password,
            };
            this.authenticated = true;
        }
        this._axios = axios.create(axiosConfig);
    }
    getFlights(beginTime, endTime) {
        const nvps = [];
        nvps.push({ begin: String(beginTime) });
        nvps.push({ end: String(endTime) });
        return this.getOpenSkyFlights(OpenSkyApi.FLIGHTS_URI, nvps);
    }
    getFlightsByArrivalAirport(airport, beginTime, endTime) {
        const nvps = [];
        nvps.push({ airport });
        nvps.push({ begin: String(beginTime) });
        nvps.push({ end: String(endTime) });
        return this.getOpenSkyFlights(OpenSkyApi.FLIGHTS_BY_ARRIVAL_URI, nvps);
    }
    getFlightsByDepartureAirport(airport, beginTime, endTime) {
        const nvps = [];
        nvps.push({ airport });
        nvps.push({ begin: String(beginTime) });
        nvps.push({ end: String(endTime) });
        return this.getOpenSkyFlights(OpenSkyApi.FLIGHTS_BY_DEPARTURE_URI, nvps);
    }
    getFlightsByAircraft(icao24, beginTime, endTime) {
        const nvps = [];
        nvps.push({ icao24 });
        nvps.push({ begin: String(beginTime) });
        nvps.push({ end: String(endTime) });
        return this.getOpenSkyFlights(OpenSkyApi.FLIGHTS_BY_AIRCRAFT_URI, nvps);
    }
    getStates(time, icao24, bbox) {
        const nvps = [];
        if (time) {
            nvps.push({ time: String(time) });
        }
        icao24?.forEach((i) => {
            nvps.push({ icao24: i });
        });
        if (bbox) {
            nvps.push({ lamin: String(bbox.minLatitude) });
            nvps.push({ lamax: String(bbox.maxLatitude) });
            nvps.push({ lomin: String(bbox.minLongitude) });
            nvps.push({ lomax: String(bbox.maxLongitude) });
        }
        if (this.checkRateLimit('GET_STATES', 4900, 9900)) {
            return this.getOpenSkyStates(OpenSkyApi.STATES_URI, nvps);
        }
        return null;
    }
    getMyStates(time, icao24, serials) {
        if (!this.authenticated) {
            throw new Error("Anonymous access of 'myStates' not allowed");
        }
        const nvps = [];
        if (time) {
            nvps.push({ time: String(time) });
        }
        icao24?.forEach((i) => {
            nvps.push({ icao24: i });
        });
        serials?.forEach((s) => {
            nvps.push({ serials: String(s) });
        });
        if (this.checkRateLimit('GET_MY_STATES', 900, 0)) {
            return this.getOpenSkyStates(OpenSkyApi.MY_STATES_URI, nvps);
        }
        return null;
    }
    async getOpenSkyStates(url, nvps) {
        const params = new URLSearchParams();
        nvps.forEach((i) => {
            for (const [key, value] of Object.entries(i)) {
                params.append(key, value);
            }
        });
        const { data } = await this._axios.get(url, {
            params,
        });
        const states = data?.states?.map((state) => stateVectorMapper(state)) || [];
        return {
            time: data.time,
            states,
        };
    }
    async getOpenSkyFlights(url, nvps) {
        const params = new URLSearchParams();
        nvps.forEach((i) => {
            for (const [key, value] of Object.entries(i)) {
                params.append(key, value);
            }
        });
        const { data } = await this._axios.get(url, {
            params,
            validateStatus: (status) => (status >= 200 && status < 300) || status === 404,
        });
        if (Array.isArray(data)) {
            return data.map((d) => flightMapper(d));
        }
        return new Array();
    }
    checkRateLimit(type, timeDiffAuth, timeDiffNoAuth) {
        const t = this.lastRequestTime[type];
        const now = Date.now();
        this.lastRequestTime[type] = now;
        return (t == null ||
            (this.authenticated && now - t > timeDiffAuth) ||
            (!this.authenticated && now - t > timeDiffNoAuth));
    }
}
export { BoundingBox } from './BoundingBox';
