// opensky-api/build/main/lib/OpenSkyApi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BoundingBox = exports.OpenSkyApi = void 0;
('use strict');
const axios_1 = __importDefault(require("axios"));
const iso_url_1 = require("iso-url");
const StateVectorMapper_1 = require("../mappers/StateVectorMapper");
const FlightMapper_1 = require("./../mappers/FlightMapper");
const axiosConfig = {
    timeout: 5000,
    headers: { 'User-Agent': 'OpenSkyApi-TS/1.0' },
};
class OpenSkyApi {
    constructor(credentials) {
        this.authenticated = false;
        this.lastRequestTime = {
            GET_STATES: null,
            GET_MY_STATES: null,
            GET_FLIGHTS: null,
        };
        if ((credentials === null || credentials === void 0 ? void 0 : credentials.username) && (credentials === null || credentials === void 0 ? void 0 : credentials.password)) {
            axiosConfig.auth = {
                username: credentials === null || credentials === void 0 ? void 0 : credentials.username,
                password: credentials === null || credentials === void 0 ? void 0 : credentials.password,
            };
            this.authenticated = true;
        }
        this._axios = axios_1.default.create(axiosConfig);
    }
    getFlights(beginTime, endTime) {
        const nvps = [];
        nvps.push({ begin: String(beginTime) });
        nvps.push({ end: String(endTime) });
        return this.getOpenSkyFlights(OpenSkyApi.FLIGHTS_URI, nvps);
    }
    getFlightsByArrivalAirport(airport, beginTime, endTime) {
        const nvps = [];
        nvps.push({ airport });
        nvps.push({ begin: String(beginTime) });
        nvps.push({ end: String(endTime) });
        return this.getOpenSkyFlights(OpenSkyApi.FLIGHTS_BY_ARRIVAL_URI, nvps);
    }
    getFlightsByDepartureAirport(airport, beginTime, endTime) {
        const nvps = [];
        nvps.push({ airport });
        nvps.push({ begin: String(beginTime) });
        nvps.push({ end: String(endTime) });
        return this.getOpenSkyFlights(OpenSkyApi.FLIGHTS_BY_DEPARTURE_URI, nvps);
    }
    getFlightsByAircraft(icao24, beginTime, endTime) {
        const nvps = [];
        nvps.push({ icao24 });
        nvps.push({ begin: String(beginTime) });
        nvps.push({ end: String(endTime) });
        return this.getOpenSkyFlights(OpenSkyApi.FLIGHTS_BY_AIRCRAFT_URI, nvps);
    }
    getStates(time, icao24, bbox) {
        const nvps = [];
        if (time) {
            nvps.push({ time: String(time) });
        }
        icao24 === null || icao24 === void 0 ? void 0 : icao24.forEach((i) => {
            nvps.push({ icao24: i });
        });
        if (bbox) {
            nvps.push({ lamin: String(bbox.minLatitude) });
            nvps.push({ lamax: String(bbox.maxLatitude) });
            nvps.push({ lomin: String(bbox.minLongitude) });
            nvps.push({ lomax: String(bbox.maxLongitude) });
        }
        if (this.checkRateLimit('GET_STATES', 4900, 9900)) {
            return this.getOpenSkyStates(OpenSkyApi.STATES_URI, nvps);
        }
        return null;
    }
    getMyStates(time, icao24, serials) {
        if (!this.authenticated) {
            throw new Error("Anonymous access of 'myStates' not allowed");
        }
        const nvps = [];
        if (time) {
            nvps.push({ time: String(time) });
        }
        icao24 === null || icao24 === void 0 ? void 0 : icao24.forEach((i) => {
            nvps.push({ icao24: i });
        });
        serials === null || serials === void 0 ? void 0 : serials.forEach((s) => {
            nvps.push({ serials: String(s) });
        });
        if (this.checkRateLimit('GET_MY_STATES', 900, 0)) {
            return this.getOpenSkyStates(OpenSkyApi.MY_STATES_URI, nvps);
        }
        return null;
    }
    async getOpenSkyStates(url, nvps) {
        var _b;
        const params = new iso_url_1.URLSearchParams();
        nvps.forEach((i) => {
            for (const [key, value] of Object.entries(i)) {
                params.append(key, value);
            }
        });
        const { data } = await this._axios.get(url, {
            params,
        });
        const states = ((_b = data === null || data === void 0 ? void 0 : data.states) === null || _b === void 0 ? void 0 : _b.map((state) => (0, StateVectorMapper_1.stateVectorMapper)(state))) || [];
        return {
            time: data.time,
            states,
        };
    }
    async getOpenSkyFlights(url, nvps) {
        const params = new iso_url_1.URLSearchParams();
        nvps.forEach((i) => {
            for (const [key, value] of Object.entries(i)) {
                params.append(key, value);
            }
        });
        const { data } = await this._axios.get(url, {
            params,
            validateStatus: (status) => (status >= 200 && status < 300) || status === 404,
        });
        if (Array.isArray(data)) {
            return data.map((d) => (0, FlightMapper_1.flightMapper)(d));
        }
        return new Array();
    }
    checkRateLimit(type, timeDiffAuth, timeDiffNoAuth) {
        const t = this.lastRequestTime[type];
        const now = Date.now();
        this.lastRequestTime[type] = now;
        return (t == null ||
            (this.authenticated && now - t > timeDiffAuth) ||
            (!this.authenticated && now - t > timeDiffNoAuth));
    }
}
exports.OpenSkyApi = OpenSkyApi;
_a = OpenSkyApi;
OpenSkyApi.HOST = 'opensky-network.org';
OpenSkyApi.API_ROOT = `https://${_a.HOST}/api`;
OpenSkyApi.STATES_URI = `${_a.API_ROOT}/states/all`;
OpenSkyApi.MY_STATES_URI = `${_a.API_ROOT}/states/own`;
OpenSkyApi.FLIGHTS_URI = `${OpenSkyApi.API_ROOT}/flights/all`;
OpenSkyApi.FLIGHTS_BY_AIRCRAFT_URI = `${OpenSkyApi.API_ROOT}/flights/aircraft`;
OpenSkyApi.FLIGHTS_BY_ARRIVAL_URI = `${OpenSkyApi.API_ROOT}/flights/arrival`;
OpenSkyApi.FLIGHTS_BY_DEPARTURE_URI = `${OpenSkyApi.API_ROOT}/flights/departure`;
var BoundingBox_1 = require("./BoundingBox");
Object.defineProperty(exports, "BoundingBox", { enumerable: true, get: function () { return BoundingBox_1.BoundingBox; } });
